File: src/leetcode/m82/RemoveDuplicatesFromSortedListII.java
/*
 * Copyright (c) 2024-2025 THGEEK.COM. All rights reserved.
 */

package leetcode.m82;

import leetcode.ListNode;

/**
 * LeetCode 82
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/10/02 08:44
 */
public class RemoveDuplicatesFromSortedListII {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        ListNode prev = dummy;
        ListNode curr = head;
        while (curr != null) {
            if (curr.next != null && curr.val == curr.next.val) {
                while (curr.next != null && curr.val == curr.next.val) {
                    curr = curr.next;
                }
                prev.next = curr.next;
            } else {
                prev = prev.next;
            }
            curr = curr.next;
        }
        return dummy.next;
    }
}

File: src/leetcode/m82/RemoveDuplicatesFromSortedListII.go
package m82

import (
	"leetcode/src/leetcode/support"
)

// LeetCode 82
func deleteDuplicates(head *support.ListNode) *support.ListNode {
	// Create a dummy node to simplify handling edge cases.
	dummy := &support.ListNode{Val: 0, Next: head}
	prev := dummy
	current := head

	for current != nil {
		// If current node is the start of a duplicate sequence...
		if current.Next != nil && current.Val == current.Next.Val {
			// Skip all nodes with the same value.
			for current.Next != nil && current.Val == current.Next.Val {
				current = current.Next
			}
			// Connect prev with the node after the duplicates.
			prev.Next = current.Next
		} else {
			// No duplicates, move prev pointer.
			prev = prev.Next
		}
		// Move current pointer forward.
		current = current.Next
	}

	return dummy.Next
}
